nemo/collections/tts/modules/transformer_dec24.py

```diff
@@ -0,0 +1,691 @@
+
+        self.norm_name = norm_name
+        if self.norm_name == 'weightnorm':
+            self.conv = nn.utils.weight_norm(self.conv)
+        elif self.norm_name == 'spectralnorm':
+            self.conv = nn.utils.spectral_norm(self.conv)
+
+    def forward(self, signal):
+        if self.is_causal:
+            padding = (int((self.kernel_size - 1) * (self.dilation)), 0)
+            signal = torch.nn.functional.pad(signal, padding)
+        conv_signal = self.conv(signal)
+        return conv_signal
+
+
+class LayerNorm(nn.Module):
+    """LayerNorm with optional bias. PyTorch doesn't support bias=False"""
+
+    def __init__(self, size, gamma0=1, eps=1e-5, use_bias=False):
+        super().__init__()
+        self.weight = nn.Parameter(torch.ones(size))
+        self.bias = nn.Parameter(torch.zeros(size)) if use_bias else None
+        self.eps = eps
+        self.size = size
+
+    def forward(self, tensor):
+        """
+        tensor (B, T, C)
+        """
+        dtype = tensor.dtype
+        # fp32 to avoid numerical issues
+        with amp.autocast(enabled=True, dtype=torch.float32):
+            tensor = F.layer_norm(tensor, self.weight.shape, self.weight, self.bias, self.eps)
+        return tensor.to(dtype)
+
+
+class LayerScale(nn.Module):
+    """
+    # borrowed from https://huggingface.co/
+    Layer scale from [Touvron et al 2021] (https://arxiv.org/abs/2103.17239).
+    Rescales diagonaly the residual outputs close to 0, with a learnt scale.
+    Args:
+        channels (int): Number of channels.
+        init (float): Initial scale.
+        channel_last (bool): True expects `[*, C]` , otherwise, `[*, C, T]`.
+        device (torch.device or None): Device to initialize the module.
+        dtype (torch.dtype or None): dtype to use to initialize the module.
+    """
+
+    def __init__(self, channels: int, init: float = 1e-4, channel_last: bool = True, device=None, dtype=None):
+        super().__init__()
+        self.channel_last = channel_last
+        self.scale = nn.Parameter(torch.full((channels,), init, requires_grad=True, device=device, dtype=dtype))
+
+    def forward(self, x: torch.Tensor):
+        dtype = x.dtype
+        with amp.autocast(enabled=True, dtype=torch.float32):
+            if self.channel_last:
+                x = self.scale * x
+            else:
+                x = self.scale[:, None] * x
+        return x.to(dtype)
+
+
+class PositionwiseConvFF(nn.Module):
+    def __init__(self, d_model, p_dropout, kernel_size=1, bias=False, is_causal=True):
+        super(PositionwiseConvFF, self).__init__()
+
+        self.d_model = d_model
+        self.non_linearity = nn.GELU(approximate="tanh")
+        self.proj = ConvNorm(d_model, d_model * 4, bias=bias, kernel_size=kernel_size, is_causal=is_causal)
+        self.o_net = ConvNorm(d_model * 4, d_model, bias=bias, kernel_size=kernel_size, is_causal=is_causal)
```

Comment:
Regarding init for these two modules: self.proj should be gpt2 style, self.o_net should be scaled by the number of layers. See https://github.com/NVIDIA/Megatron-LM/blob/d4e72c0d33edc0c53aeb624f617eb77cebce6ae9/megatron/core/transformer/mlp.py#L66-L92 where the defaults are described at https://github.com/NVIDIA/Megatron-LM/blob/d4e72c0d33edc0c53aeb624f617eb77cebce6ae9/megatron/core/transformer/transformer_config.py#L125

These are recommended for MLPs, once we move to convs, I don't know if it makes sense to continue with the same init scheme.

Reply:
This is hardcoded in during Transformer init